crates/core/app/tests/view_server_can_be_served_on_localhost.rs

```diff
@@ -0,0 +1,205 @@
+use {
+    self::common::BuilderExt,
+    anyhow::Context,
+    cnidarium::TempStorage,
+    penumbra_app::{
+        genesis::{self, AppState},
+        server::consensus::Consensus,
+    },
+    penumbra_asset::STAKING_TOKEN_ASSET_ID,
+    penumbra_keys::{keys::AddressIndex, test_keys},
+    penumbra_mock_client::MockClient,
+    penumbra_mock_consensus::TestNode,
+    penumbra_proto::{
+        view::v1::{
+            view_service_client::ViewServiceClient, view_service_server::ViewServiceServer,
+            StatusRequest, StatusResponse,
+        },
+        DomainType,
+    },
+    penumbra_view::{Planner, SpendableNoteRecord, ViewClient},
+    std::ops::Deref,
+    tap::{Tap, TapFallible},
+};
+
+mod common;
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+async fn view_server_can_be_served_on_localhost() -> anyhow::Result<()> {
+    // Install a test logger, acquire some temporary storage, and start the test node.
+    let guard = common::set_tracing_subscriber();
+    let storage = TempStorage::new().await?;
+
+    // Instantiate a mock tendermint proxy, which we will connect to the test node.
+    let proxy = penumbra_mock_tendermint_proxy::TestNodeProxy::new::<Consensus>();
+
+    // Start the test node.
+    let mut test_node = {
+        let app_state = AppState::Content(
+            genesis::Content::default().with_chain_id(TestNode::<()>::CHAIN_ID.to_string()),
+        );
+        let consensus = Consensus::new(storage.as_ref().clone());
+        TestNode::builder()
+            .single_validator()
+            .with_penumbra_auto_app_state(app_state)?
+            .on_block(proxy.on_block_callback())
+            .init_chain(consensus)
+            .await
+            .tap_ok(|e| tracing::info!(hash = %e.last_app_hash_hex(), "finished init chain"))?
+    };
+
+    // Sync the mock client, using the test wallet's spend key, to the latest snapshot.
+    let mut client = MockClient::new(test_keys::SPEND_KEY.clone())
+        .with_sync_to_storage(&storage)
+        .await?
+        .tap(
+            |c| tracing::info!(client.notes = %c.notes.len(), "mock client synced to test storage"),
+        );
+
+    // Jump ahead a few blocks.
+    test_node
+        .fast_forward(10)
+        .tap(|_| tracing::debug!("fast forwarding past genesis"))
+        .await?;
+
+    let grpc_url = "http://127.0.0.1:8080"
+        .parse::<url::Url>()?
+        .tap(|url| tracing::debug!(%url, "parsed grpc url"));
```

Comment:
Can we avoid using a network socket here, and use an entirely in-memory GRPC service (the way `pcli` does a `BoxGrpcService`, for instance), so that we can run all of these tests concurrently without contending for the same global environment state?

Reply:
i would very much like to do this too. i have pulled on this thread and found that it's going to be a slightly involved change. the assumption that the rpc service lives across a network connection is baked into a few different places in the `ViewServer` and its underlying `Worker`.

the notion of a remote url is threaded around in places like `record_block`, here: https://github.com/penumbra-zone/penumbra/blob/main/crates/view/src/storage.rs#L1366-L1373
..and indirectly via `tonic::transport::Channel` in a few other places.

i'll take a swing at this, and report back.

Reply:
this is doable, and i have most of that change assembled now.

there is a final hurdle, `Sync` bounds, that i'll handle next. for now, the relevant incremental commits are:

* 4e43181ef - view: use type-erased BoxGrpcService (does not build)
* 9fea1cf12 - view: `ViewServer` holds a `Channel`
* 44eb082c0 - view: hoist url out of worker
* 4cb63624b - view: avoid passing url to `record_block()`

Reply:
@hdevalence this has turned into a bit of a quagmire. `BoxGrpcService` isn't `Sync`, but `ViewService` has a `Sync` bound. i've tried various approaches, but am finding this to be an invasive change.

particularly with some other items on my docket related to the testnet faucet, i think a prudent course of action for us is to:
* revert 4e43181, and punt on using an in-memory service **for now.** the changes in 4cb6362, 44eb082, and 9fea1cf all are good stepping stones we can keep.
* file a ticket to track view server support for an in-memory grpc service in the view server.
* pivot to other high-priority mainnet tickets, and revisit the in-memory grpc service after that work winds down.

what do you think?

Reply:
oh, and to disambiguate: this test _does_ use an in-memory **view** service for the view client, but we do not have a way for the view service to talk to an in-memory rpc endpoint (_is there a better word for the broader rpc router returned by `penumbra_app::rpc::router()` served by `pd`?_).

Reply:
@hdevalence and i talked a bit directly and agreed that in-memory support for the `ViewServer`'s rpc endpoint can be punted on for now. i've dropped the two tail commits off this branch, and it should be ready for a review now.

i will mark this as resolved, now that i've filed #4517 to track this